Lib/importlib/_bootstrap_external.py

```diff
@@ -1685,6 +1688,19 @@ def _fill_cache(self):
         if sys.platform.startswith(_CASE_INSENSITIVE_PLATFORMS):
             self._relaxed_path_cache = {fn.lower() for fn in contents}
 
+        self._cache_is_normalized = False
+
+    def _normalize_cache(self):
+        """
+        Normalize all entries in the caches to NFKC.
+        """
+        from unicodedata import normalize
+
+        self._path_cache = { normalize('NFKC', p) for p in self._path_cache }
+        self._relaxed_path_cache = { normalize('NFKC', p) for p in self._relaxed_path_cache }
```

Comment:
AFAICT, the cache does not influence the filename that gets returned.  As long as that filename matches what's in the cache, which is effectively a superset of allowed paths, we will get correct behavior.  That's because the code in this loader's `find_spec()` checks to make sure the actual filename (or dirname for packages) exists.

For the sake of `__import__()` and `importlib.import_module()`, we may need to preserve the un-normalized entries in the cache:

```suggestion
        self._path_cache.update(normalize('NFKC', p) for p in self._path_cache)
        self._relaxed_path_cache.update(normalize('NFKC', p) for p in self._relaxed_path_cache)
```

That's because those functions do not normalize the module name like happens for import statements.  Thus a module that currently imports via those functions could start raising `ModuleNotFoundError` because the unnormalized name isn't in the cache any more.

Reply:
There's another problem here that the PR doesn't solve: the normalized filename might not exist.  The normalized cache means the module might make it past that point.  However, the subsequent `_path_isfile(full_path)` (or `_path_isdir(base_path)`) will return `False`, leading to the same `ModuleNotFoundError` we see now.

For example:

```python
import os
name = '\u03BCmod'  # normalizes to '\u00B5'
filename = name + '.py'
with open(filename, 'w') as outfile:
    outfile.write('SPAM = 42')
ns = {}
try:
    exec(f'import {name}', ns, ns)
finally:
    os.unlink(filename)
```

Without this PR the above script fails with `ModuleNotFoundError`.  With this PR, it still fails (I'm pretty sure).  This is the same issue that PEP 3131 caused (also discussed in gh-55161).

One solution is to fall back to the known file for the relevant `_path_isfile()` and `_path_isdir()` calls.  We would store the actual path here:

```suggestion
        self._path_cache = { normalize('NFKC', p) for p in self._path_cache }
        self._relaxed_path_cache = { normalize('NFKC', p) for p in self._relaxed_path_cache }
```

(`_fill_cache()` would have to be updated to make `_path_cache` a dict with values set to `None`.)

Further, it might make sense to keep the normalized entries separate, like we do for "relaxed":

```suggestion
        self._normalized_path_cache = { normalize('NFKC', p): p for p in self._path_cache }
        self._normalized_relaxed_path_cache = { normalize('NFKC', p): p for p in self._relaxed_path_cache }
```

Then we could drop `self._cache_is_normalized` in lieu of `self._normalized_cache_path is not None`.  The fallback code in `find_spec()` would get the known actual filename from the cache and use that in place of directly using the provided module name.

That would even help somewhat with the case where there are multiple files that normalize to the same name.

Reply:
There are probably some test cases we could add to cover some of these corner cases.

Reply:
> There's another problem here that the PR doesn't solve: the normalized filename might not exist. The normalized cache means the module might make it past that point. However, the subsequent _path_isfile(full_path) (or _path_isdir(base_path)) will return False, leading to the same ModuleNotFoundError we see now.

Thanks, I've been thinking about why tests pass on macOS and fail on Linux and this is likely it: Linux filesystems are not unicode-aware whereas APFS on macOS performs normalisation aware comparisons.

Reply:
Yeah, I suspected something along those lines.

Reply:
Resolved by keeping both the normalized and unnormalised names around, see https://github.com/python/cpython/pull/113447#discussion_r1438171918.